src/main/kotlin/org/jitsi/utils/stats/BucketStats.kt

```diff
@@ -0,0 +1,137 @@
+    init {
+        if (!thresholdsNoMax.contentEquals(thresholdsNoMax.sortedArray())) {
+            throw IllegalArgumentException("Thresholds must be sorted: ${thresholdsNoMax.joinToString()}")
+        }
+    }
+    private val totalValue = LongAdder()
+    private val totalCount = LongAdder()
+    private val average: Double
+        get() = totalValue.sum() / totalCount.sum().toDouble()
+    private val maxValue = AtomicLong(0)
+    private val buckets = Buckets(thresholdsNoMax)
+
+    fun addValue(value: Long) {
+        if (value >= 0) {
+            totalValue.add(value)
+            maxValue.maxAssign(value)
+            totalCount.increment()
+            buckets.addValue(value)
+        }
+    }
+
+    fun toJson() = OrderedJsonObject().apply {
+        val snapshot = snapshot
+        put("average$averageMaxLabel", snapshot.average)
+        put("max$averageMaxLabel", snapshot.maxValue)
+        put("total_count", snapshot.totalCount)
+
+        put("buckets", getBucketsJson(snapshot.buckets))
+    }
+
+    val snapshot: Snapshot
+        get() = Snapshot(
+            average = average,
+            maxValue = maxValue.get(),
+            totalCount = totalCount.sum(),
+            buckets = buckets.snapshot
+        )
+
+    data class Snapshot(
+        val average: Double,
+        val maxValue: Long,
+        val totalCount: Long,
+        val buckets: Buckets.Snapshot
+    )
+
+    fun getBucketsJson(b: Buckets.Snapshot) = OrderedJsonObject().apply {
+        for (i in 0..b.buckets.size - 2) {
+            put("<= ${b.buckets[i].first}$bucketLabel", b.buckets[i].second)
+        }
+
+        val indexOfSecondToLast = b.buckets.size - 2
+        put("> ${b.buckets[indexOfSecondToLast].first}$bucketLabel", b.buckets.last().second)
+
+        put("p99<=", b.p99bound)
+        put("p999<=", b.p999bound)
+    }
+}
+
+class Buckets(thresholdsNoMax: LongArray) {
+    private val thresholds = longArrayOf(*thresholdsNoMax, Long.MAX_VALUE)
+    private val thresholdCounts = Array(thresholds.size + 1) { LongAdder() }
+    val snapshot: Snapshot
+        get() {
+            val bucketCounts = Array(thresholds.size) { i -> Pair(thresholds[i], thresholdCounts[i].sum()) }
+
+            var p99 = Long.MAX_VALUE
+            var p999 = Long.MAX_VALUE
+            var sum: Long = 0
+            val totalCount = bucketCounts.map { it.second }.sum()
+            bucketCounts.forEach {
+                sum += it.second
+                if (it.first < p99 && sum > 0.99 * totalCount) p99 = it.first
+                if (it.first < p999 && sum > 0.999 * totalCount) p999 = it.first
+            }
+
+            // Not enough data
+            if (totalCount < 100 || p99 == Long.MAX_VALUE) p99 = -1
+            if (totalCount < 1000 || p999 == Long.MAX_VALUE) p999 = -1
+            return Snapshot(bucketCounts, p99, p999)
+        }
+
+    private fun findBucket(value: Long): LongAdder {
```

Comment:
Would you care to make this `protected open` in case anyone wants to optimize for a different distribution (just based on the comment below)?